### ISSUE TITLE
Proposal: Result-based API

### ISSUE THREAD

**Issue by jsha** [algesten/ureq#128]
Right now, ureq handles errors by `.synthetic_error()` (for protocol-level errors, like connection refused), and `.error()`, (for protocol-level errors *and* HTTP-semantic errors like 404 and 500).

It would be very convenient to use Results for errors, in particular because it allows using the `?` sugar to trivially propagate errors.

I think we can offer a new API without breaking backwards compatibility (for now). We should define a new method on Request like so:

```
fn do(&self) -> Result<Response, Error>
```

By default, this method would return Error only for protocol-level errors; for 4xx and 5xx, the user would be responsible for checking the status code and reading the body. However, for convenience we could add another method on Request:

```
fn error_on_status(&mut self) -> &mut Request
```

When the user sets this option for a request, 4xx and 5xx errors have their bodies read automatically and get turned into `Error`s.

One other change needs to happen as part of this: send_bytes, send_string, send_form, and send_json currently return Response. To provide equivalents in the Result-based API, I propose we define `body_bytes`, `body_string`, `body_form`, and `body_json`, which set the body of a Request object to one of those things, so the body gets sent as part of `.do()`. Example:

```
  let resp = ureq::post("https://myapi.example.com/ingest")
      .set("X-My-Header", "Secret")
      .error_on_status()
      .body_json(json!({
          "name": "martin",
          "rust": true
      }))
      .do()?;
  println!("{}", resp.into_string()?);
```

Related: #94 #125

Thoughts? @BurntSushi @algesten

**Comment by BurntSushi** [comment]
I _think_ this looks pretty good to me, although it does seem a little bit of a bummer that you need to duplicate the `send_*` methods.

The other API that stands out to me is adding something on `Response`. That would make the existing `Request -> Response` still correct. So e.g., you might have `ureq::get(&url).call().error_for_status()?`. Although perhaps `error_for_status` isn't quite the right name. I think the verbose but correct name is `error_for_synthetic_or_status`, so we would just need to pick a better name.

Perhaps you could just add `into_error` and `into_synthetic_error` APIs on `Response`. I guess they would both return a `Result<Response, ureq::Error>`. The former would return an error for either a synthetic error or for a 4xx/5xx response, while the latter would only return an error for a synthetic error. That would also resolve #126 too I think. Other possible names might be `into_result` and `into_synthetic_result`.

I don't have much experience with HTTP implementations though, so I don't know what the right choice is. But this option sticks out to me as one that's a bit more cohesive with the _current_ API at least. I think it requires fewer changes.

**Comment by jsha** [comment]
> Perhaps you could just add into_error and into_synthetic_error APIs on Response.

This would work, but I prefer to add a method that returns a Result, with the intent that it eventually become the main way people use the library. I think it's more natural for people writing Rust these days and makes it easier to ensure that errors get handled (or ignored on purpose).

> it does seem a little bit of a bummer that you need to duplicate the send_* methods.

A couple of other possibilities here:

 - Don't duplicate these, but just make the whole thing as a breaking change with a major version bump; change everything that returns Response to return Result<Response, Error>.
 - Don't duplicate them just yet, and just offer the `.do()` method. The `send_*` methods are fairly thin wrappers, and users can readily replace them by serializing JSON, or a form, in their own code (and setting the appropriate Content-Type).

**Comment by algesten** [comment]
Synthetic errors is my biggest regret in ureq, and I just want to be rid of them. Maybe that clouds my view :) 

My preferred route would be to keep the API clean (not doubling up on body methods) and embrace the breaking change. I would start in the end of _how would I want to use it_?

I think it's a reasonable default is that 4xx and 5xx does surface as `Error` in a common error path, but we provide ways to remove that behavior.

`Request::redirects(42)` allows setting `0` to get `3xx` response codes in the `Response`. I suggest we do similar new functions for 4xx and 5xx. Something like this (names totally debatable):

`Request::error_on_4xx(false)`
`Request::error_on_5xx(false)`

That means for a consume of ureq, the change from 1.x to 2.x would be adding `?`, and potentially removing some code around `Response::synthetic_error`.

```rust
  let resp = ureq::post("https://myapi.example.com/ingest")
      .set("X-My-Header", "Secret")
      .send_json(json!({
          "name": "martin",
          "rust": true
      }))?;  // New ? since all errors surface here  

  println!("{}", resp.into_string()?);
```

**Comment by jsha** [comment]
I made an update on https://github.com/algesten/ureq/pull/132#issuecomment-683558034 with something similar to what you mention in the last comment here, though I haven't allowed separate disablement of 4xx errors vs 5xx errors. I think that overcomplicates things, and clients can match on the error's Response anyhow.

This still doesn't solve for one of the problems we were concerned about: Making the common path for errors include reading the body, so clients reuse connections. One thought on an independent change for that: We could decide that all Responses (error or not) automatically read the first 2000 bytes into an internal buffer, so long as the response sets Content-Length.

**Comment by algesten** [comment]
I like it! A couple of questions on the PR https://github.com/algesten/ureq/pull/132#commitcomment-41902979

Reading 2000 bytes sounds like a good idea. I think we could do that both for content-length and transfer-encoding chunked, or do you a problem with that?

**Comment by jsha** [comment]
@BurntSushi @sergiimk @jyn514 We've published a release candidate of 2.0.0-rc2: https://crates.io/crates/ureq/2.0.0-rc2. That includes the change to use a Result instead of synthetic errors, along with a bunch of other refactorings: https://github.com/algesten/ureq/blob/main/CHANGELOG.md. Since you all had expressed interest in improving the error ergonomics in ureq, I would be very grateful if you had time to give it a try and confirm whether it meets your needs. Thanks!

**Comment by jyn514** [comment]
@jsha thanks, this is really exciting! Unfortunately, the new Error can't be introspected at all, which is really limiting - `kind()` is crate-local, so the only thing you can do with an error is call `to_string()` on it.

Deadlinks wants to know in particular whether an error came from failing to connect to the server (no internet, invalid URL syntax, etc.) or whether the connection succeeded and the server returned a non-2XX status: https://github.com/deadlinks/cargo-deadlinks/blob/96cb6cb3bc04bf05cf8a656b903dc1f73f32b33c/src/check.rs#L247-L261

**Comment by jsha** [comment]
Good point. we should export at least `kind()`. Are there other things you need to inspect from Error?

**Comment by jyn514** [comment]
Hmm, so I actually misread the docs - `ureq::Error` is *only* for failing to connect to the server, and non-2xx responses are part of the `Response`. In that case the existing API is probably ok (although `kind()` would still be nice) - let me play around with it some more.

**Comment by jsha** [comment]
Actually, you read right. By default, non-2xx responses will return a `ureq::Error` with `kind == HTTP`. However, you can turn off that behavior with `error_on_non_2xx(false)`. If you turn off the behavior, you get non-2xx responses as a normal Response.

**Comment by jyn514** [comment]
Hmm, that seems a little confusing - maybe instead there could be `Error::into_http_error` which returns `Ok(Response)` if it was non-2xx and `Err(Self)` otherwise? Then all the `error()` functions make sense; as-is, they'll never be used unless you set `error_on_non_2xx`.

**Comment by BurntSushi** [comment]
Yeah, I do like the `into_http_error` idea better, defined on the response for the same reasons as what @jyn514 said. It does tweak the default a bit, but I actually feel like that might be what folks expect? I think HTTP libraries generally return responses regardless of their status code. Turning non-2xx/3xx responses into an actual error is usually a separate step after that. The `into_http_error` idea preserves that separate step while making it convenient.

But I'm not sure. Maybe I'm wrong about most HTTP libraries, or maybe most HTTP libraries are themselves wrong and non-2xx/3xx should be handled as errors _by default_.

**Comment by jyn514** [comment]
Hmm, that's different from the idea I had actually - I was suggesting that error_on_non2xx always be on, your saying it should always be off (and have a separate function to turn responses into errors).

**Comment by BurntSushi** [comment]
If it's always on, how do you get a `Response` with a non-2xx/3xx status? Maybe I'm just missing something.

**Comment by jyn514** [comment]
It's returned as an error: https://github.com/algesten/ureq/pull/257

**Comment by BurntSushi** [comment]
Interesting. Well, I defer to y'all. I don't have much experience with HTTP libraries other than my impression that your proposed default is opposite of what other HTTP libraries do. (I'm thinking of `requests` in Python in particular.)

**Comment by jsha** [comment]
@BurntSushi, I agree, most HTTP libraries do the opposite of the path we're taking: The user code has to explicitly decide what to do with 4xx and 5xx statuses. For instance, [Go's net/http says](https://golang.org/pkg/net/http/#Client.Do): "A non-2xx status code doesn't cause an error." I actually don't 100% know why they behave that way. I suspect it's because in many languages it's hard to unwrap an error type to get at the underlying value. It may also be because reading a response body is a potentially-expensive step. Libraries usually leave it up to the user to decide when and how to read the response body, making it hard to decide what to do in the error case. In our case, we don't process the body on errors, but soon (#264) will process it if it has already been read into our 8k buffer.

My initial inclination was the same as yours - better to behave as other HTTP libraries do, and require explicit opt-in from the user to turn status codes into error. However, I've been convinced that the error-by-default approach @algesten and @jyn514 favor is better. My experience as a user of other libraries has usually been that I'm frustrated by the need to explicitly handle 4xx and 5xx. I often forget to handle them altogether. If I do handle them, I have to add back in some of the niceties that the library normally handles, like including the URL that caused the error. [Here's one anecdotal post](http://www.metabates.com/2015/10/15/handling-http-request-errors-in-go/.) suggesting that failure to handle 4xx and 5xx is a common problem in Go programs.

By the way, in #257, here's the approach we decided on:

```
pub enum Error {
    Status(u16, Response),
    Transport(Transport),
}
```

So you can extract an error Response using a match statement. It's unconventional to be sure, but I think it will turn out to be very pleasant to work with.

**Comment by BurntSushi** [comment]
@jsha Sounds good to me! If nothing else, it will prove to be a great experiment. :-) Thank you very much for explaining!

**Comment by jsha** [comment]
Moving conversation over from https://github.com/algesten/ureq/pull/281#issuecomment-751524883:

@jsha:
> > I'm slightly reluctant to add this on the grounds that it adds multiple ways to do the same thing - you could do the enum matching style, or you could do into_response. That's not the worst thing, though, and it seems like some folks may want this style. I guess it's a way of hedging our bets on the unusual API choice of making non-2xx errors by default.

@algesten:
> Maybe we could do one more double take on that decision before finalizing 2.0? If it genuinely is so surprising to users, maybe it's a bad choice? I dunno. I want to make the most user friendly, least surprising lib…

Sure! Happy to think some more on this. Previously in this issue I [talked about some of the reasons](https://github.com/algesten/ureq/issues/128#issuecomment-739568229) I'm guessing other libraries don't treat Status errors as errors by default: hard to unwrap an error to get a Response; reading body potentially expensive. I think we have some good answers for those.

There are a few ways we could go about making this decision more methodically:  We could post in some Rust forums and see what people think of the proposed API. Or we could review a sample of extant code that uses various popular HTTP clients to see whether users more commonly want to treat non-2xx as errors or not. I suspect we may actually find cases where non-2xx was simply not handled because only the happy path got tested.

(related: #262)

**Comment by jsha** [comment]
Here are some thoughts from Brad Fitzpatrick, who has maintained Go's net/http package for 10 years now:

https://github.com/bradfitz/exp-httpclient/blob/master/problems.md#too-easy-to-not-check-return-status-codes
https://github.com/golang/go/issues/23707
https://github.com/bradfitz/exp-httpclient

In short, he suggests it would be better to treat non-2xx as an error by default. For me, that carries a lot of weight.

**Comment by algesten** [comment]
This does indeed carry a lot of weight. With that I'll put my doubts to rest.

Thanks!

**Comment by jyn514** [comment]
This was implemented in 2.0, so I think it can be closed :)
